File: backend/app/api/v1/schemas/erp.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ClientCreate(BaseModel):
    nif: str | None = None
    name: str
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    city: str | None = None
    postal_code: str | None = None
    client_type: str = "customer"
    marketing_consent: bool = False


class ClientResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    nif: str | None = None
    name: str
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    city: str | None = None
    postal_code: str | None = None
    client_type: str
    marketing_consent: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Nuevos Modelos de Facturación Avanzada ---


class ProductCreate(BaseModel):
    item_type: str = "product"  # product | service
    sku: str | None = None
    barcode: str | None = None
    name: str
    description: str | None = None
    category: str | None = None
    location: str | None = None
    unit: str = "ud"
    price: float = 0.0
    cost_price: float | None = None
    tax_percentage: float = 21.0
    stock_quantity: int = 0
    stock_min_alert: int = 0
    is_active: bool = True
    supplier_id: UUID | None = None
    reorder_quantity: int | None = None


class ProductResponse(ProductCreate):
    id: UUID
    tenant_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InvoiceLineCreate(BaseModel):
    product_id: UUID | None = None
    description: str
    quantity: float = 1.0
    unit_price: float = 0.0
    discount_percentage: float = 0.0
    tax_percentage: float = 21.0
    # No pedimos el 'total' al frontend, lo calcularemos nosotros si queremos mayor seguridad
    # o bien podemos aceptar el total como sugerencia y re-verificarlo.


class InvoiceLineResponse(InvoiceLineCreate):
    id: UUID
    invoice_id: UUID
    total: float

    model_config = ConfigDict(from_attributes=True)


class ClientUpdate(BaseModel):
    nif: str | None = None
    name: str | None = None
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    city: str | None = None
    postal_code: str | None = None
    client_type: str | None = None
    marketing_consent: bool | None = None


class ProductUpdate(BaseModel):
    item_type: str | None = None
    sku: str | None = None
    barcode: str | None = None
    name: str | None = None
    description: str | None = None
    category: str | None = None
    location: str | None = None
    unit: str | None = None
    price: float | None = None
    cost_price: float | None = None
    tax_percentage: float | None = None
    stock_quantity: int | None = None
    stock_min_alert: int | None = None
    is_active: bool | None = None
    supplier_id: UUID | None = None
    reorder_quantity: int | None = None


class InvoiceStatusUpdate(BaseModel):
    status: str


class InvoiceCreate(BaseModel):
    serie: str | None = "F"  # Serie de facturación: F=normal, R=rectificativa, T=simplificada
    invoice_number: str | None = None  # Si se provee se ignora la numeración automática
    date: datetime
    due_date: datetime | None = None
    status: str = "draft"
    invoice_type: str = "issued"
    notes: str | None = None
    terms: str | None = None
    # Régimen fiscal especial 303: intracomunitario | isp | recargo_equivalencia
    fiscal_regime: str | None = None
    # Retención IRPF Art.95 (facturas recibidas de profesionales)
    retencion_irpf_rate: float | None = None
    retencion_irpf_amount: float | None = None
    lines: list[InvoiceLineCreate] = []


class InvoiceResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    client_id: UUID
    invoice_number: str | None = None
    date: datetime
    due_date: datetime | None = None
    amount_base: float
    tax_amount: float
    amount_total: float
    status: str
    invoice_type: str
    rectifies_invoice_id: UUID | None = None
    rectification_reason: str | None = None
    fiscal_regime: str | None = None
    retencion_irpf_rate: float | None = None
    retencion_irpf_amount: float | None = None
    notes: str | None = None
    terms: str | None = None
    external_id: str | None = None
    created_at: datetime
    updated_at: datetime

    client: ClientResponse | None = None
    lines: list[InvoiceLineResponse] = []

    model_config = ConfigDict(from_attributes=True)


# --- Stock / Inventario ---


class StockMovementCreate(BaseModel):
    movement_type: str  # entrada | salida | ajuste
    quantity: int
    unit_cost: float | None = None
    reference: str | None = None
    notes: str | None = None


class StockMovementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    product_id: UUID
    user_id: UUID | None = None
    movement_type: str
    quantity: int
    stock_after: int
    unit_cost: float | None = None
    reference: str | None = None
    notes: str | None = None
    created_at: datetime


class StockValuationByCategory(BaseModel):
    category: str | None = None
    units: int
    value: float
    product_count: int


class StockValuationResponse(BaseModel):
    total_value: float
    total_units: int
    product_count: int
    missing_cost_price_count: int
    by_category: list[StockValuationByCategory]


# --- Pedidos de Venta ---


class SalesOrderLineCreate(BaseModel):
    product_id: UUID | None = None
    description: str
    quantity: float = 1.0
    unit_price: float = 0.0
    discount_percentage: float = 0.0
    tax_percentage: float = 21.0


class SalesOrderLineResponse(SalesOrderLineCreate):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    order_id: UUID
    total: float


class SalesOrderCreate(BaseModel):
    client_id: UUID
    order_number: str | None = None
    date: datetime | None = None
    expected_delivery: datetime | None = None
    notes: str | None = None
    quote_id: UUID | None = None
    lines: list[SalesOrderLineCreate] = []


class SalesOrderUpdate(BaseModel):
    status: str | None = None
    expected_delivery: datetime | None = None
    notes: str | None = None


class SalesOrderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    client_id: UUID
    order_number: str | None = None
    date: datetime
    expected_delivery: datetime | None = None
    status: str
    amount_base: float
    tax_amount: float
    amount_total: float
    notes: str | None = None
    quote_id: UUID | None = None
    created_at: datetime
    updated_at: datetime

    client: ClientResponse | None = None
    lines: list[SalesOrderLineResponse] = []


# --- Pedidos de Compra ---


class PurchaseOrderLineCreate(BaseModel):
    product_id: UUID | None = None
    description: str
    quantity: float = 1.0
    unit_price: float = 0.0
    tax_percentage: float = 21.0


class PurchaseOrderLineResponse(PurchaseOrderLineCreate):
    received_quantity: float = 0

    model_config = ConfigDict(from_attributes=True)
    id: UUID
    order_id: UUID
    total: float


class PurchaseOrderCreate(BaseModel):
    supplier_id: UUID
    order_number: str | None = None
    date: datetime | None = None
    expected_delivery: datetime | None = None
    notes: str | None = None
    lines: list[PurchaseOrderLineCreate] = []


class PurchaseOrderUpdate(BaseModel):
    status: str | None = None
    expected_delivery: datetime | None = None
    notes: str | None = None


class PurchaseOrderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tenant_id: UUID
    supplier_id: UUID
    order_number: str | None = None
    date: datetime
    expected_delivery: datetime | None = None
    status: str
    amount_base: float
    tax_amount: float
    amount_total: float
    notes: str | None = None
    created_at: datetime
    updated_at: datetime
    supplier: ClientResponse | None = None
    lines: list[PurchaseOrderLineResponse] = []


# --- Facturación Recurrente ---

from datetime import date as DateType


class RecurringLineItem(BaseModel):
    description: str
    quantity: float = 1.0
    unit_price: float = 0.0
    tax_percentage: float = 21.0


class RecurringInvoiceCreate(BaseModel):
    client_id: UUID
    name: str
    interval_type: str = "monthly"  # monthly | quarterly | yearly | weekly
    next_run_date: DateType
    notes: str | None = None
    terms: str | None = None
    lines: list[RecurringLineItem] = []


class RecurringInvoiceUpdate(BaseModel):
    name: str | None = None
    interval_type: str | None = None
    next_run_date: DateType | None = None
    is_active: bool | None = None
    notes: str | None = None
    terms: str | None = None
    lines: list[RecurringLineItem] | None = None


class RecurringInvoiceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    tenant_id: UUID
    client_id: UUID
    name: str
    interval_type: str
    next_run_date: DateType
    last_run_date: DateType | None = None
    is_active: bool
    lines_json: list[dict]
    notes: str | None = None
    terms: str | None = None
    created_at: datetime
    updated_at: datetime
    client: ClientResponse | None = None

File: frontend/src/lib/api/erp.ts
import { request, downloadBlob, requestUpload } from "./client";

export interface ScannedInvoiceLine {
    description: string;
    quantity: number;
    unit_price: number;
    tax_percentage: number;
    total: number;
}

export interface ScannedInvoice {
    emisor: {
        nif: string;
        name: string;
        address: string | null;
        city: string | null;
        postal_code: string | null;
    };
    invoice_number: string;
    issue_date: string;
    due_date: string | null;
    lines: ScannedInvoiceLine[];
    amount_base: number;
    tax_amount: number;
    amount_total: number;
    currency: string;
    payment_method: string | null;
    iban: string | null;
    confidence: number;
    warnings: string[];
}

export interface Client {
    id: string;
    nif: string | null;
    name: string;
    email: string | null;
    phone: string | null;
    address: string | null;
    city: string | null;
    postal_code: string | null;
    client_type: string;
    created_at: string;
    updated_at: string | null;
}

export interface InvoiceLine {
    id?: string;
    product_id?: string | null;
    description: string;
    quantity: number;
    unit_price: number;
    discount_percentage: number;
    tax_percentage: number;
    total?: number;
}

export interface Invoice {
    id: string;
    client_id: string;
    invoice_number: string | null;
    date: string;
    due_date: string | null;
    amount_base: number;
    tax_amount: number;
    amount_total: number;
    status: string;
    invoice_type: string;
    notes: string | null;
    terms: string | null;
    external_id: string | null;
    verifactu_status: string | null;
    verifactu_sent_at: string | null;
    created_at: string;
    updated_at: string | null;
    client?: Client;
    lines?: InvoiceLine[];
}

export interface Product {
    id: string;
    tenant_id: string;
    item_type: string;
    sku: string | null;
    barcode: string | null;
    name: string;
    description: string | null;
    category: string | null;
    location: string | null;
    unit: string;
    price: number;
    cost_price: number | null;
    tax_percentage: number;
    stock_quantity: number;
    stock_min_alert: number;
    is_active: boolean;
    created_at: string;
    updated_at: string | null;
}

export interface StockMovement {
    id: string;
    tenant_id: string;
    product_id: string;
    user_id: string | null;
    movement_type: string;
    quantity: number;
    stock_after: number;
    unit_cost: number | null;
    reference: string | null;
    notes: string | null;
    created_at: string;
}

export interface StockValuationByCategory {
    category: string | null;
    units: number;
    value: number;
    product_count: number;
}

export interface StockValuation {
    total_value: number;
    total_units: number;
    product_count: number;
    missing_cost_price_count: number;
    by_category: StockValuationByCategory[];
}

export interface QuoteLine {
    id?: string;
    product_id?: string | null;
    description: string;
    quantity: number;
    unit_price: number;
    tax_percentage: number;
    total_line?: number;
}

export interface Quote {
    id: string;
    client_id: string;
    quote_number: string | null;
    date: string;
    valid_until: string | null;
    amount_base: number;
    tax_amount: number;
    amount_total: number;
    status: string; // draft | sent | accepted | rejected
    notes: string | null;
    terms: string | null;
    opportunity_id: string | null;
    created_at: string;
    client?: Client;
    lines?: QuoteLine[];
}

export interface SalesOrderLine {
    id?: string;
    product_id?: string | null;
    description: string;
    quantity: number;
    unit_price: number;
    discount_percentage: number;
    tax_percentage: number;
    total?: number;
}

export interface SalesOrder {
    id: string;
    tenant_id: string;
    client_id: string;
    order_number: string | null;
    date: string;
    expected_delivery: string | null;
    status: string;
    amount_base: number;
    tax_amount: number;
    amount_total: number;
    notes: string | null;
    quote_id: string | null;
    created_at: string;
    updated_at: string;
    client?: Client;
    lines?: SalesOrderLine[];
}

export interface PurchaseOrderLine {
    id?: string;
    product_id?: string | null;
    description: string;
    quantity: number;
    unit_price: number;
    tax_percentage: number;
    total?: number;
}

export interface PurchaseOrder {
    id: string;
    tenant_id: string;
    supplier_id: string;
    order_number: string | null;
    date: string;
    expected_delivery: string | null;
    status: string;
    amount_base: number;
    tax_amount: number;
    amount_total: number;
    notes: string | null;
    created_at: string;
    updated_at: string;
    supplier?: Client;
    lines?: PurchaseOrderLine[];
}

export interface RecurringLineItem {
    description: string;
    quantity: number;
    unit_price: number;
    tax_percentage: number;
}

export interface RecurringInvoice {
    id: string;
    tenant_id: string;
    client_id: string;
    name: string;
    interval_type: string;
    next_run_date: string;
    last_run_date: string | null;
    is_active: boolean;
    lines_json: RecurringLineItem[];
    notes: string | null;
    terms: string | null;
    created_at: string;
    updated_at: string;
    client?: Client;
}

export const erp = {
    clients: {
        list: (params?: { skip?: number; limit?: number; client_type?: string }) => {
            const q = new URLSearchParams(params as Record<string, string>).toString();
            return request<Client[]>(`/api/v1/clients${q ? "?" + q : ""}`);
        },
        create: (data: Partial<Client>) =>
            request<Client>("/api/v1/clients", {
                method: "POST",
                body: JSON.stringify(data),
            }),
        update: (id: string, data: Partial<Client>) =>
            request<Client>(`/api/v1/clients/${id}`, {
                method: "PATCH",
                body: JSON.stringify(data),
            }),
        delete: (id: string) =>
            request<void>(`/api/v1/clients/${id}`, { method: "DELETE" }),
        invoices: (id: string, params?: { skip?: number; limit?: number }) => {
            const q = new URLSearchParams(params as Record<string, string>).toString();
            return request<Invoice[]>(`/api/v1/clients/${id}/invoices${q ? "?" + q : ""}`);
        },
        createInvoice: (clientId: string, data: Partial<Invoice>) =>
            request<Invoice>(`/api/v1/clients/${clientId}/invoices`, {
                method: "POST",
                body: JSON.stringify(data),
            }),
    },
    products: {
        list: (params?: {
            skip?: number;
            limit?: number;
            q?: string;
            category?: string;
            is_active?: boolean;
            status?: "ok" | "low_stock" | "out_of_stock";
        }) => {
            const filtered: Record<string, string> = {};
            if (params) {
                for (const [k, v] of Object.entries(params)) {
                    if (v !== undefined && v !== null && v !== "") {
                        filtered[k] = String(v);
                    }
                }
            }
            const qs = new URLSearchParams(filtered).toString();
            return request<Product[]>(`/api/v1/products${qs ? "?" + qs : ""}`);
        },
        byBarcode: (code: string) =>
            request<Product>(`/api/v1/products/by-barcode/${encodeURIComponent(code)}`),
        create: (data: Partial<Product>) =>
            request<Product>("/api/v1/products", {
                method: "POST",
                body: JSON.stringify(data),
            }),
        update: (id: string, data: Partial<Product>) =>
            request<Product>(`/api/v1/products/${id}`, {
                method: "PATCH",
                body: JSON.stringify(data),
            }),
        delete: (id: string) =>
            request<void>(`/api/v1/products/${id}`, { method: "DELETE" }),
    },
    invoices: {
        list: (params?: { skip?: number; limit?: number }) => {
            const q = new URLSearchParams(params as Record<string, string>).toString();
            return request<Invoice[]>(`/api/v1/invoices${q ? "?" + q : ""}`);
        },
        get: (id: string) => request<Invoice>(`/api/v1/invoices/${id}`),
        create: (clientId: string, data: Partial<Invoice>) =>
            request<Invoice>(`/api/v1/clients/${clientId}/invoices`, {
                method: "POST",
                body: JSON.stringify(data),
            }),
        updateStatus: (id: string, status: string) =>
            request<Invoice>(`/api/v1/invoices/${id}/status`, {
                method: "PATCH",
                body: JSON.stringify({ status }),
            }),
        delete: (id: string) =>
            request<void>(`/api/v1/invoices/${id}`, { method: "DELETE" }),
        scan: (file: File) => {
            const fd = new FormData();
            fd.append("file", file);
            return requestUpload<ScannedInvoice>("/api/v1/invoices/scan", fd);
        },
        /** Aprendizaje OCR (F2.5): el usuario corrigió una extracción → guardar
         * overrides para que las próximas facturas del mismo NIF salgan
         * mejor sin coste extra de LLM. */
        learnScanCorrection: (params: {
            supplier_nif: string;
            original: ScannedInvoice;
            corrected: ScannedInvoice;
        }) =>
            request<{ saved: boolean; overrides: Record<string, unknown> }>(
                "/api/v1/invoices/scan/learn",
                { method: "POST", body: JSON.stringify(params) },
            ),
        downloadPdf: (id: string, invoiceNumber: string | null) =>
            downloadBlob(`/api/v1/invoices/${id}/pdf`, `Factura_${invoiceNumber || id.slice(0, 8)}.pdf`),
        downloadFacturae: (id: string, invoiceNumber: string | null) =>
            downloadBlob(`/api/v1/invoices/${id}/facturae`, `facturae_${(invoiceNumber || id.slice(0, 8)).replace(/\//g, "-")}.xsig`),
        sendVerifactu: (id: string) =>
            request<{ message: string; invoice_id: string; status: string }>(`/api/v1/invoices/${id}/verifactu-send`, { method: "POST" }),
    },
    quotes: {
        list: (params?: { skip?: number; limit?: number }) => {
            const q = new URLSearchParams(params as Record<string, string>).toString();
            return request<Quote[]>(`/api/v1/quotes${q ? "?" + q : ""}`);
        },
        create: (data: Partial<Quote>) =>
            request<Quote>("/api/v1/quotes", {
                method: "POST",
                body: JSON.stringify(data),
            }),
        update: (id: string, data: Partial<Quote>) =>
            request<Quote>(`/api/v1/quotes/${id}`, {
                method: "PATCH",
                body: JSON.stringify(data),
            }),
        convertToInvoice: (id: string) =>
            request<{ invoice_id: string; invoice_number: string; amount_total: number; message: string }>(
                `/api/v1/quotes/${id}/convert-to-invoice`,
                { method: "POST" }
            ),
        delete: (id: string) => request<void>(`/api/v1/quotes/${id}`, { method: "DELETE" }),
    },
    orders: {
        list: () => request<SalesOrder[]>("/api/v1/orders"),
        create: (data: Partial<SalesOrder> & { lines?: SalesOrderLine[] }) =>
            request<SalesOrder>("/api/v1/orders", { method: "POST", body: JSON.stringify(data) }),
        update: (id: string, data: Partial<SalesOrder>) =>
            request<SalesOrder>(`/api/v1/orders/${id}`, { method: "PATCH", body: JSON.stringify(data) }),
        delete: (id: string) => request<void>(`/api/v1/orders/${id}`, { method: "DELETE" }),
    },
    purchaseOrders: {
        list: () => request<PurchaseOrder[]>("/api/v1/purchase-orders"),
        create: (data: Partial<PurchaseOrder> & { lines?: PurchaseOrderLine[] }) =>
            request<PurchaseOrder>("/api/v1/purchase-orders", { method: "POST", body: JSON.stringify(data) }),
        update: (id: string, data: Partial<PurchaseOrder>) =>
            request<PurchaseOrder>(`/api/v1/purchase-orders/${id}`, { method: "PATCH", body: JSON.stringify(data) }),
        delete: (id: string) => request<void>(`/api/v1/purchase-orders/${id}`, { method: "DELETE" }),
    },
    recurring: {
        list: () => request<RecurringInvoice[]>("/api/v1/recurring-invoices"),
        create: (data: Partial<RecurringInvoice> & { lines?: RecurringLineItem[] }) =>
            request<RecurringInvoice>("/api/v1/recurring-invoices", { method: "POST", body: JSON.stringify(data) }),
        update: (id: string, data: Partial<RecurringInvoice> & { lines?: RecurringLineItem[] }) =>
            request<RecurringInvoice>(`/api/v1/recurring-invoices/${id}`, { method: "PATCH", body: JSON.stringify(data) }),
        delete: (id: string) => request<void>(`/api/v1/recurring-invoices/${id}`, { method: "DELETE" }),
        run: (id: string) => request<Invoice>(`/api/v1/recurring-invoices/${id}/run`, { method: "POST" }),
    },
    stock: {
        valuation: () => request<StockValuation>("/api/v1/stock/valuation"),
        movements: (productId: string) => request<StockMovement[]>(`/api/v1/products/${productId}/stock-movements`),
        addMovement: (productId: string, data: Partial<StockMovement>) =>
            request<StockMovement>(`/api/v1/products/${productId}/stock-movements`, {
                method: "POST",
                body: JSON.stringify(data),
            }),
    },
};
